### PR TITLE
Introduce snapshot/restore high level API

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request defines the reference traits `Snapshotable` and `Transportable` as a first step. Once the base is defined, it introduces the user API (OpenAPI and CLI) to snapshot and restore a VM.
For now, snapshot and restore are both empty actions.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -231,6 +231,14 @@ fn create_app<'a, 'b>(
                 .default_value(&api_server_path)
                 .group("vmm-config"),
         )
+        .arg(
+            Arg::with_name("restore-from")
```
Comment (sameo) [inline src/main.rs:235]: I tested that one (I did not before, shamelessly...) and it works.
I'd prefer if we call it `--restore` though, sorry for the nitpicking.

File: src/main.rs [hunk 2]
```diff
@@ -231,6 +231,14 @@ fn create_app<'a, 'b>(
...
+                .long("restore-from")
+                .help("Restore from a VM snapshot.")
```
Comment (sameo) [inline src/main.rs:237]: I think here we should document that the argument for this should be a valid URL, e.g. `file:///foo/bar/` or tcp://192.168.1.10/`

**Comment by sboeuf** [comment]
@sameo 
>Should we add the ch-remote support through that PR?

Yes I totally forgot but I'm working on it right now.

File: src/main.rs (hunk 1 above)
Comment (sboeuf) [reply src/main.rs:235]: Oh yeah no problem, and of course I tested it too :)

**Comment by sboeuf** [comment]
@sameo this is ready now, PTAL

**Comment by sameo** [comment]
> @sameo this is ready now, PTAL

LGTM

File: vm-migration/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    /// Add a SnapshotDatasection to the component snapshot data.
+    pub fn add_data_section(&mut self, section: SnapshotDataSection) {
+        self.snapshot_data.insert(section.id.clone(), section);
+    }
+}
+
+/// A snapshotable component can be snapshoted.
+pub trait Snapshotable: Pausable {
```
Comment (rbradford) [inline vm-migration/src/lib.rs:107]: s/Snapshotable/Snapshottable/ so you don't make native English speakers go crazy.

File: vmm/src/api/openapi/cloud-hypervisor.yaml [hunk 4]
```diff
@@ -171,6 +171,24 @@ paths:
...
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/VmSnapshotConfig'
+        required: true
+      responses:
+        204:
+          description: The VM instance was successfully snapshotted.
+        404:
+          description: The VM instance could not be snapshotted because it is not created.
```
Comment (rbradford) [inline vmm/src/api/openapi/cloud-hypervisor.yaml]: You got it right here!

File: vmm/src/migration.rs [hunk 5]
```diff
@@ -59,3 +62,25 @@ pub fn recv_vm_snapshot(source_url: &str) -> std::result::Result<Snapshot, Migra
...
+    if let Some(vm_section) = snapshot
+        .snapshot_data
+        .get(&format!("{}-section", VM_SNAPSHOT_ID))
+    {
+        let vm_snapshot: VmSnapshot =
+            serde_json::from_slice(&vm_section.snapshot).map_err(|e| {
+                MigratableError::Restore(anyhow!("Could not deserialize VM snapshot {}", e))
+            })?;
+
+        //        self.set_state(&ioapic_state);
```
Comment (rbradford) [inline vmm/src/migration.rs:78]: Strange comment here?

Comment (sameo) [reply vmm/src/migration.rs:78]: Crufts...This needs to be removed indeed.

File: vm-migration/src/lib.rs (hunk 3 above)
Comment (sboeuf) [reply vm-migration/src/lib.rs:107]: Ok let me change that :)

**Comment by sboeuf** [comment]
@rbradford I addressed the comments! PTAL

**Pull request merged**